src/couch_file.erl

```diff
@@ -223,29 +223,53 @@ delete(RootDir, Filepath) ->
 
 
 delete(RootDir, Filepath, Async) ->
-    DelFile = filename:join([RootDir,".delete", ?b2l(couch_uuids:random())]),
+    DelFile = deleted_filename(RootDir, Filepath),
+    RenameOnDelete = config:get_boolean("couchdb", "rename_on_delete", false),
     case file:rename(Filepath, DelFile) of
-    ok ->
-        if (Async) ->
+        ok when RenameOnDelete ->
+            Now = calendar:local_time(),
+            file:change_time(DelFile, Now);
+        ok when Async ->
             spawn(file, delete, [DelFile]),
             ok;
-        true ->
-            file:delete(DelFile)
-        end;
-    Error ->
-        Error
+        ok ->
+            file:delete(DelFile);
+        Error ->
+            Error
     end.
 
+deleted_filename(RootDir, Filepath) ->
+    RootName = filename:rootname(Filepath),
+    DelFileRoot = couch_util:url_encode(RootName -- RootDir),
+    Suffix = deleted_filename_suffix(),
+    Extention = filename:extension(Filepath),
+    DelFileName = io_lib:format("~s.~s~s", [DelFileRoot, Suffix, Extention]),
+    filename:join([RootDir,".delete", DelFileName]).
+
+deleted_filename_suffix() ->
+    {{Y, Mon, D}, {H, Min, S}} = calendar:universal_time(),
+    io_lib:format("~w~2.10.0B~2.10.0B"
+        ".~2.10.0B~2.10.0B~2.10.0B.~6s",
+        [Y, Mon, D, H, Min, S, couch_uuids:random()]).
 
 nuke_dir(RootDelDir, Dir) ->
+    case config:get_boolean("couchdb", "rename_on_delete", false) of
+        true -> couch_server:delete_file(RootDelDir, Dir, []);
```

Comment:
I don't like the fact that we introduce cyclic dependency between couch_file and couch_server here. But I wouldn't insist on changing it. Feel free to ignore.